internal/pkg/v2/infrastructure/redis/reading.go

```diff
@@ -0,0 +1,77 @@
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package redis
+
+import (
+	"encoding/json"
+
+	"github.com/edgexfoundry/edgex-go/internal/pkg/common"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+
+	model "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/google/uuid"
+)
+
+const ReadingsCollection = "v2:reading"
+
+var emptyBinaryValue = make([]byte, 0)
+
+// Add a reading to the database
+func addReading(conn redis.Conn, r model.Reading) (reading model.Reading, err error) {
+	var m []byte
+	var baseReading *model.BaseReading
+	switch newReading := r.(type) {
```

Comment:
Can now use r.ValueType to determine if Binary or Simple, rather than reflection. Always best to avoid reflection if possible.

Reply:
@lenny-intel Lenny, If changing to not use type conversion, the code will look like below
```
	if r.GetBaseReading().ValueType == dtos.ValueTypeBinary {
		// Handle BinaryReading value
		// Clear the binary data since we do not want to persist binary data to save on memory.
		newReading := r.(model.BinaryReading)
		newReading.BinaryValue = emptyBinaryValue

		baseReading = &newReading.BaseReading
		if err = checkReadingValue(baseReading); err != nil {
			return nil, err
		}
		m, err = json.Marshal(newReading)
		reading = newReading
	} else {
		// Handle SimpleReading value
		newReading := r.(model.SimpleReading)
		baseReading = &newReading.BaseReading
		if err = checkReadingValue(baseReading); err != nil {
			return nil, err
		}
		m, err = json.Marshal(newReading)
		reading = newReading
	}
```
Will still need to do the type conversion inside the if/else condition and which might cause panic when the `ValueType` is invalid.  Seems that use the original code might be a safer way. Any other suggestion on this? Thanks.

Reply:
Somewhere before we get here, we should be validating that `ValueType` is set appropriately and not make it here with an invalid type. I think we should be validating value of `ValueType` here: https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/reading.go#L64

Reply:
Also, I get your point on the type conversion... ;-) Going to have it ether way, so I am good with the switch/case approach.